From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: cuckm <cuckmeron@gmail.com>
Date: Sun, 20 Oct 2024 10:53:13 -0400
Subject: [PATCH] MineZ

(cherry picked from commit 34a204d49796aa0f27b937b4ddb3fed666f475ed)

diff --git a/net/minecraft/data/worldgen/DimensionTypes.java b/net/minecraft/data/worldgen/DimensionTypes.java
index 1356ae93cca0a9e16112f61b599eec7d2ea7efd7..49f7e5d812b27c1a0f3494d6bbd2c7ac437d08ca 100644
--- a/net/minecraft/data/worldgen/DimensionTypes.java
+++ b/net/minecraft/data/worldgen/DimensionTypes.java
@@ -26,7 +26,7 @@ public class DimensionTypes {
                 BlockTags.INFINIBURN_OVERWORLD,
                 BuiltinDimensionTypes.OVERWORLD_EFFECTS,
                 0.0F,
-                new DimensionType.MonsterSettings(false, true, UniformInt.of(0, 7), 0)
+                new DimensionType.MonsterSettings(false, true, ConstantInt.of(15), 15)//0, 7), 0)
             )
         );
         context.register(
@@ -86,7 +86,7 @@ public class DimensionTypes {
                 BlockTags.INFINIBURN_OVERWORLD,
                 BuiltinDimensionTypes.OVERWORLD_EFFECTS,
                 0.0F,
-                new DimensionType.MonsterSettings(false, true, UniformInt.of(0, 7), 0)
+                new DimensionType.MonsterSettings(false, true, ConstantInt.of(15), 15)//0, 7), 0)
             )
         );
     }
diff --git a/net/minecraft/world/effect/RegenerationMobEffect.java b/net/minecraft/world/effect/RegenerationMobEffect.java
index 81481267a1577721dcc405f39a4c350bd59ac9a2..295edd95f92583218fa5c766f79f8fe20e74f0c2 100644
--- a/net/minecraft/world/effect/RegenerationMobEffect.java
+++ b/net/minecraft/world/effect/RegenerationMobEffect.java
@@ -19,7 +19,7 @@ class RegenerationMobEffect extends MobEffect {
 
     @Override
     public boolean shouldApplyEffectTickThisTick(int duration, int amplifier) {
-        int i = 50 >> amplifier;
+        int i = 25 >> amplifier; // MineZ 50
         return i <= 0 || duration % i == 0;
     }
 }
